osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/misc/GetGsmDiagnosticCommandExecutor.java

```diff
@@ -0,0 +1,312 @@
+
+  private static final int RESULT_OPERATOR_INDEX = 0;
+  private static final int RESULT_MODEM_REGISTRATION_STATUS_INDEX = 1;
+  private static final int RESULT_CIRCUIT_SWITCHED_STATUS_INDEX = 2;
+  private static final int RESULT_PACKET_SWITCHED_STATUS_INDEX = 3;
+  private static final int RESULT_CELL_INFO_INDEX = 4;
+  private static final int RESULT_ADJACENT_CELLS_INDEX = 5;
+  private static final int RESULT_CAPTURE_TIME_INDEX = 6;
+
+  private static final int CELL_INFO_CELL_ID_INDEX = 0;
+  private static final int CELL_INFO_LOCATION_ID_INDEX = 1;
+  private static final int CELL_INFO_SIGNAL_QUALITY_INDEX = 2;
+  private static final int CELL_INFO_BIT_ERROR_RATE_INDEX = 3;
+  private static final int CELL_INFO_MOBILE_COUNTRY_CODE_INDEX = 4;
+  private static final int CELL_INFO_MOBILE_NETWORK_CODE_INDEX = 5;
+  private static final int CELL_INFO_CHANNEL_NUMBER_INDEX = 6;
+
+  private static final int ADJACENT_CELLS_CELL_ID_INDEX = 0;
+  private static final int ADJACENT_CELLS_SIGNAL_QUALITY_INDEX = 1;
+
+  private final DlmsHelper dlmsHelper;
+  private final DlmsObjectConfigService dlmsObjectConfigService;
+
+  @Autowired
+  public GetGsmDiagnosticCommandExecutor(
+      final DlmsHelper dlmsHelper, final DlmsObjectConfigService dlmsObjectConfigService) {
+    super(GetGsmDiagnosticRequestDto.class);
+
+    this.dlmsHelper = dlmsHelper;
+    this.dlmsObjectConfigService = dlmsObjectConfigService;
+  }
+
+  @Override
+  public GetGsmDiagnosticRequestDto fromBundleRequestInput(final ActionRequestDto bundleInput)
+      throws ProtocolAdapterException {
+
+    this.checkActionRequestType(bundleInput);
+
+    return new GetGsmDiagnosticRequestDto();
+  }
+
+  @Override
+  public GetGsmDiagnosticResponseDto execute(
+      final DlmsConnectionManager conn,
+      final DlmsDevice device,
+      final GetGsmDiagnosticRequestDto getGsmDiagnosticQuery)
+      throws ProtocolAdapterException {
+
+    final DlmsObject dlmsObject =
+        this.dlmsObjectConfigService.findDlmsObjectForCommunicationMethod(
+            device, DlmsObjectType.GSM_DIAGNOSTIC);
+
+    final AttributeAddress[] addresses = this.createAttributeAddresses(dlmsObject);
+
+    final String addressesDescriptions = JdlmsObjectToStringUtil.describeAttributes(addresses);
+
+    conn.getDlmsMessageListener()
+        .setDescription("Get GsmDiagnostic, retrieve attributes: " + addressesDescriptions);
+
+    LOGGER.info("Get GsmDiagnostic, retrieve attributes: {}", addressesDescriptions);
+
+    final List<GetResult> getResultList =
+        this.dlmsHelper.getAndCheck(conn, device, "Get GsmDiagnostic", addresses);
+
+    final String resultString =
+        getResultList.stream().map(this::resultToString).collect(Collectors.joining("-", "{", "}"));
+    LOGGER.info("GetResultList: {}", resultString);
+
+    if (getResultList.stream()
+        .noneMatch(result -> result.getResultCode() == AccessResultCode.SUCCESS)) {
```

Comment:
This could happen when get-with-list is not supported, then each attribute is requested seperately. The idea was to return as much as possible and the rest of GXF and the SMHE should be able to handle missing values.

Reply:
Hmm, maybe this deserves some further discussion outside of this pull request.
I would say that attempts to return partially constructed results are less desirable than to just return failure if something goes wrong. I don't believe such strategy is applied in any of the earlier built operations either.

One objection I can think of is that this kind of implementation complicates the API. The decision whether that's a good idea or not should be made with a broader perspective than that for a single client application of the platform.
When you get a successfull result you no longer know this for sure until after you have validated it somehow, and you would need to make elements in the response optional that in a normal return value might be considered required.
It would also make it harder I guess to gain insight into success/failure rates for the platform requests, as you would need to inspect results to see if they need to be included (partially?) in failure counts.

Reply:
Changed to the suggested check to keep it consistent with the rest of the code.